// ss-comparator/dist/index.esm.js
/**
 * enumeration of compare result, there is only three state
 *
 * @export
 * @enum {number}
 */
var CompareResult;
(function (CompareResult) {
    CompareResult[CompareResult["EQUAL"] = 0] = "EQUAL";
    CompareResult[CompareResult["GREATER"] = 1] = "GREATER";
    CompareResult[CompareResult["LESS"] = -1] = "LESS";
})(CompareResult || (CompareResult = {}));
/**
 * a class that describe how compares two object
 *
 * @export
 * @class Comparator
 */
var Comparator = /** @class */ (function () {
    /**
     * Creates an instance of Comparator.
     * @param {compareFunction} compareFunction - function that implement compare operation
     * @memberof Comparator
     */
    function Comparator(compareFunction) {
        this.compare = compareFunction || Comparator.defaultCompareFunction;
    }
    /**
     * @param {(string|number)} a - compare target a
     * @param {(string|number)} b - compare target b
     * @returns {number}
     */
    Comparator.defaultCompareFunction = function (a, b) {
        if (a === b) {
            return CompareResult.EQUAL;
        }
        return a < b ? CompareResult.LESS : CompareResult.GREATER;
    };
    /**
     * compare if equal or not, a === b
     *
     * @param {*} a
     * @param {*} b
     * @returns {boolean}
     * @memberof Comparator
     */
    Comparator.prototype.equal = function (a, b) {
        return this.compare(a, b) === CompareResult.EQUAL;
    };
    /**
     * compare if smaller. a < b
     *
     * @param {*} a
     * @param {*} b
     * @returns {boolean}
     * @memberof Comparator
     */
    Comparator.prototype.lessThan = function (a, b) {
        return this.compare(a, b) === CompareResult.LESS;
    };
    /**
     * compare if greater. a > b
     *
     * @param {*} a
     * @param {*} b
     * @returns {boolean}
     * @memberof Comparator
     */
    Comparator.prototype.greaterThan = function (a, b) {
        return this.compare(a, b) === CompareResult.GREATER;
    };
    /**
     * compare if smaller or equal. a <= b
     *
     * @param {*} a
     * @param {*} b
     * @returns {boolean}
     * @memberof Comparator
     */
    Comparator.prototype.lessThanOrEqual = function (a, b) {
        return this.lessThan(a, b) || this.equal(a, b);
    };
    /**
     * compare if greater or equal. a >= b
     *
     * @param {*} a
     * @param {*} b
     * @returns {boolean}
     * @memberof Comparator
     */
    Comparator.prototype.greaterThanOrEqual = function (a, b) {
        return this.greaterThan(a, b) || this.equal(a, b);
    };
    /**
     * reverse the compare function
     *
     * @memberof Comparator
     */
    Comparator.prototype.reverse = function () {
        var compareOriginal = this.compare;
        this.compare = function (a, b) { return compareOriginal(b, a); };
    };
    return Comparator;
}());

export default Comparator;
export { CompareResult };
